docs/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,111 @@
+
+cuSpatial has two main components,
+the cuSpatial-python package and libcuspatial c++ package.
+This article dicusses the design of cuspatial-python,
+for information of libcuspatial,
+see [libcuspatial developer guide](link) and [c++ API reference](link).
+We refer to the python package as `cuspatial` and the c++ package as `libcuspatial` below.
+
+## Overview
+
+At a high level, `cuspatial` has three parts:
+- A GPU backed `GeoDataFrame` data structure
+- A set of computation APIs
+- A cython API layer
+
+## GPU Accelerated `GeoDataFrame` and `GeoSeries`
+
+Note: since the core data structure of cuspatial shares the same name as that
+of `geopandas`.
+We refer to geopandas' dataframe object as `geopandas.GeoDataFrame` and to
+cuspatial's dataframe object as `GeoDataFrame`.
+
+Under the hood,
+cuspatial can perform parallel computation on geometry data thanks to its [structure of array](https://en.wikipedia.org/wiki/Parallel_array) (SoA) format.
+Specifically,
+cuspatial adopts geoarrow format as the SoA standard.
+Geoarrow is a derived data type from the arrow list type adopting a [`Variable-size List Layout`](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout),
+with the inner-most layer storing the points with a `Fixed-size list layout` array with `size=2`.
+Per definition,
+each increase in geometry complexity (dimension, or multi-geometry) requires an extra level of indirection.
+Geoarrow allows a mixture of geometry types to present in the same column by adopting the [Dense Union Array Layout](https://arrow.apache.org/docs/format/Columnar.html#dense-union).
+In cusptial,
+we refer to each level of indirection from highest level to lowest by
+`geometries`, `parts`, `rings` and `coordinates`.
+The first three are integral offset arrays and the last is an floating-point interleaved xy-coordinated array.
+Read [geoarrow format documentation](https://github.com/geopandas/geo-arrow-spec/blob/main/format.md) specification for more detail.
+
+Cuspatial implements partial arrow union array via `GeoColumn` and `GeoMeta`.
+A GeoColumn is a composition of child columns and a `GeoMeta` object.
+The `GeoMeta` owns two arrays that are similar to the types buffer and offsets buffer from dense union.
+
+```{note} Currently,
+`GeoColumn` only implements four concrete array types: `points`, `multipoints`, `multilinestrings` (named as lines) and `multipolygon` (named as polygons).
+Linestrings and multilinestrings are stored uniformly as multilinestrings in the multilinestrings array.
+Polygons and multipolygons are stored uniformly as multipolygons in multipolygon array.
+Points and multipoints are stored separately in different arrays.
+This is because storing points in a multipoints array requires 50% more storage overhead.
+While this may also be true for linestrings and polygons,
+previous use cases of cuspatial often involves more complex linestrings and polygons,
+and the storage overhead is lower compared to points.
+```
```

Comment:
```suggestion
```{note}
Currently, `GeoColumn` only implements four concrete array types: `points`,
`multipoints`, multilinestrings (called `lines`) and multipolygons (called
`polygons`). Linestrings and multilinestrings are stored uniformly as
multilinestrings in the `multilinestrings` array. Polygons and multipolygons are
stored uniformly as multipolygons in the `multipolygons` array.

Points and multipoints are stored separately in different arrays, because
storing points in a multipoints array requires 50% more storage overhead.
While this may also be true for linestrings and polygons, many uses of
cuSpatial involve more complex linestrings and polygons, where the 
storage overhead of multigeometry indirection is lower compared to points.
```
```